<a href="https://colab.research.google.com/github/NEEL5252/Global-Data-Analysis/blob/main/Analyzing_of_the_global_terrorism_dataset.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **PROJECT NAME**: Global Terrorism Dataset

### **PROJECT TYPE:** EDA
### **CONTRIBUTOR:** Individual

# **PROJECT SUMMARY**

Using the Global Terrorism Database (GTD) from 1970 to 2017, this research investigates essential facets of global terrorism through extensive analysis. The findings include locations with various assault intensities, trends in annual attack frequency, success rates, and the top nations with both the highest and lowest fatalities. It dives into the number of injuries received in each nation, the number of fatalities and injuries per area, the weapons used, and the geographical predominance of assault-based assaults. This study strengthens policymaking, counterterrorism efforts, and academic research by providing greater insights into the varied processes of terrorism. Finally, it prepares stakeholders to improve global security measures, creating a proactive approach to effectively battling terrorism.


# GITHUB link:


https://github.com/NEEL5252

# Problem Statement:

1. Attacks overall by areas
2. Total assaults for the year, whether they went down or up.
3. The number of successful and unsuccessful assaults.
4. The top 10 nations in terms of both kill rates and kills per capita
5. The total number of victims injured in terrorist acts worldwide
6. The overall death toll by geographic region
7. The overall number of victims per location of injuries
8. The total number of weapons used
9. The location of all the assaults that took place.


# Buisseness Objective

# **LET'S START THE ANALYSIS**

## **PHASE - 1**


```
#  Importing libraries, importing dataset and read the dataset
```



In [155]:
# Importing necessary libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import plotly.express as px
import seaborn as sns
import plotly.graph_objects as go

In [156]:
# Mount the google drive
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [213]:
# Load the csv file / Read the csv file
globalTd = pd.read_csv("/content/drive/MyDrive/Global Terrarisom dataSet/Global Terrorism Data.csv", encoding='ISO-8859-1')

<ipython-input-213-6f2e09c616af>:2: DtypeWarning:

Columns (4,6,31,33,61,62,63,76,79,90,92,94,96,114,115,121) have mixed types. Specify dtype option on import or set low_memory=False.



In [158]:
# Print the globalTd data
globalTd.head()

eventid  iyear  imonth  iday approxdate  extended resolution  country  \
0  197000000001   1970       7     2        NaN         0        NaN       58   
1  197000000002   1970       0     0        NaN         0        NaN      130   
2  197001000001   1970       1     0        NaN         0        NaN      160   
3  197001000002   1970       1     0        NaN         0        NaN       78   
4  197001000003   1970       1     0        NaN         0        NaN      101   

          country_txt  region  ... addnotes scite1 scite2  scite3  dbsource  \
0  Dominican Republic       2  ...      NaN    NaN    NaN     NaN      PGIS   
1              Mexico       1  ...      NaN    NaN    NaN     NaN      PGIS   
2         Philippines       5  ...      NaN    NaN    NaN     NaN      PGIS   
3              Greece       8  ...      NaN    NaN    NaN     NaN      PGIS   
4               Japan       4  ...      NaN    NaN    NaN     NaN      PGIS   

   INT_LOG  INT_IDEO INT_MISC INT_ANY  related  
0        0         0        0       0      NaN  
1        0         1        1       1      NaN  
2       -9        -9        1       1      NaN  
3       -9        -9        1       1      NaN  
4       -9        -9        1       1      NaN  

[5 rows x 135 columns]

In [159]:
# Shape of the globalTd dataset
globalTd.shape

(181691, 135)

In [160]:
# Count the null and missing values from the particular column of the dataset
nullcounts = globalTd.isnull().sum()
# Total null and missing counts from the dataset
totalNull = nullcounts.sum()
print(totalNull)

13853997


### Insights of the dataset:-


Together, there are 181691 rows and columns. 1,38,53,997 pieces of information are either missing or null in total. The year, kind, quantity of fatalities, nature of the weapon used, quantity of injuries, and city of the assault are among the details included in this data collection. Null values and missing values for several columns need to be removed. The wrangling and cleansing of the data will soon be finished.



## **PHASE - 2:-**


```
# Understanding the dataset
```



In [161]:
# Describing our dataset
globalTd.describe()

eventid          iyear         imonth           iday  \
count  1.816910e+05  181691.000000  181691.000000  181691.000000   
mean   2.002705e+11    2002.638997       6.467277      15.505644   
std    1.325957e+09      13.259430       3.388303       8.814045   
min    1.970000e+11    1970.000000       0.000000       0.000000   
25%    1.991021e+11    1991.000000       4.000000       8.000000   
50%    2.009022e+11    2009.000000       6.000000      15.000000   
75%    2.014081e+11    2014.000000       9.000000      23.000000   
max    2.017123e+11    2017.000000      12.000000      31.000000   

            extended        country         region       latitude  \
count  181691.000000  181691.000000  181691.000000  177135.000000   
mean        0.045346     131.968501       7.160938      23.498343   
std         0.208063     112.414535       2.933408      18.569242   
min         0.000000       4.000000       1.000000     -53.154613   
25%         0.000000      78.000000       5.000000      11.510046   
50%         0.000000      98.000000       6.000000      31.467463   
75%         0.000000     160.000000      10.000000      34.685087   
max         1.000000    1004.000000      12.000000      74.633553   

          longitude    specificity  ...     ransomamt   ransomamtus  \
count  1.771340e+05  181685.000000  ...  1.350000e+03  5.630000e+02   
mean  -4.586957e+02       1.451452  ...  3.172530e+06  5.784865e+05   
std    2.047790e+05       0.995430  ...  3.021157e+07  7.077924e+06   
min   -8.618590e+07       1.000000  ... -9.900000e+01 -9.900000e+01   
25%    4.545640e+00       1.000000  ...  0.000000e+00  0.000000e+00   
50%    4.324651e+01       1.000000  ...  1.500000e+04  0.000000e+00   
75%    6.871033e+01       1.000000  ...  4.000000e+05  0.000000e+00   
max    1.793667e+02       5.000000  ...  1.000000e+09  1.320000e+08   

         ransompaid  ransompaidus  hostkidoutcome     nreleased  \
count  7.740000e+02    552.000000    10991.000000  10400.000000   
mean   7.179437e+05    240.378623        4.629242    -29.018269   
std    1.014392e+07   2940.967293        2.035360     65.720119   
min   -9.900000e+01    -99.000000        1.000000    -99.000000   
25%   -9.900000e+01      0.000000        2.000000    -99.000000   
50%    0.000000e+00      0.000000        4.000000      0.000000   
75%    1.273412e+03      0.000000        7.000000      1.000000   
max    2.750000e+08  48000.000000        7.000000   2769.000000   

             INT_LOG       INT_IDEO       INT_MISC        INT_ANY  
count  181691.000000  181691.000000  181691.000000  181691.000000  
mean       -4.543731      -4.464398       0.090010      -3.945952  
std         4.543547       4.637152       0.568457       4.691325  
min        -9.000000      -9.000000      -9.000000      -9.000000  
25%        -9.000000      -9.000000       0.000000      -9.000000  
50%        -9.000000      -9.000000       0.000000       0.000000  
75%         0.000000       0.000000       0.000000       0.000000  
max         1.000000       1.000000       1.000000       1.000000  

[8 rows x 77 columns]

In [162]:
# Calculating total number of kills and injured people during terrorist attacks
globalTd[['nkill', 'nwound']].describe()

nkill         nwound
count  171378.000000  165380.000000
mean        2.403272       3.167668
std        11.545741      35.949392
min         0.000000       0.000000
25%         0.000000       0.000000
50%         0.000000       0.000000
75%         2.000000       2.000000
max      1570.000000    8191.000000

In [163]:
# Counting total success attacks
globalTd[globalTd['success'] == 1]['success'].sum()

161632

In [164]:
# Different types of weapons used in terrorist attacks
globalTd['weaptype1_txt'].unique()

array(['Unknown', 'Explosives', 'Incendiary', 'Firearms', 'Chemical',
       'Melee', 'Sabotage Equipment',
       'Vehicle (not to include vehicle-borne explosives, i.e., car or truck bombs)',
       'Fake Weapons', 'Radiological', 'Other', 'Biological'],
      dtype=object)

In [165]:
# Types of the attacks
globalTd['attacktype1_txt'].unique()

array(['Assassination', 'Hostage Taking (Kidnapping)',
       'Bombing/Explosion', 'Facility/Infrastructure Attack',
       'Armed Assault', 'Hijacking', 'Unknown', 'Unarmed Assault',
       'Hostage Taking (Barricade Incident)'], dtype=object)

### Description

From the above description:
1. Total 181691 terrorist attacks have been occured till 2017
2. In latitude and longitude we have around 4000 null values which shows that the we don't know the attacked area
3. Total 171378 people were lost their lives during terrorist attacks
4. Total 165380 people were injured during terrorist attacks
5. There are many columns which we don't need we will remove them in near future during the data wrangling phase
6. Out of 181691, 161632 attacks have been successfully executed and remains were unsuccessfull attacks
7. There are different types of weapons were used in terrorist attacks i.e. Explosives, firearms, chemical etc...
8. There are many different types of attacks were done in the past like Assassination, bombing, hijacking, kidnapping etc...

## **PHASE - 3**


```
Data Wrangling
```


1. Remove columns which will not used in our analyis, which contains more numbers of null and missing values

In [166]:
# Make list of columns which we want to delete
deleteColumns = []
for i in globalTd.columns:
  deleteColumns.append(i)

In [167]:
# columns list which are not gonna deleted (Based on these columns we will analysis the dataset)
notDeleteColumns = ['evenid', 'iyear', 'country', 'country_txt', 'region', 'region_txt',
                    'provstate', 'city', 'latitude', 'longitude', 'location',
                    'doubtterr', 'alternative', 'multiple','success',
                    'suicide', 'attacktype1', 'attacktype1_txt', 'targtype1',
                    'targtype1_txt', 'targsubtype1', 'targsubtype1_txt', 'corp1',
                    'weaptype1', 'weaptype1_txt', 'nkill', 'nwound']

# Converting it into array
notDeleteColumnsArray = np.array(notDeleteColumns)

In [168]:
# Deleting not necessary columns from the dataset
for i in deleteColumns:
  if (notDeleteColumnsArray[notDeleteColumnsArray == i]):
    pass
  else:
    globalTd.drop(i, inplace=True, axis=1)

<ipython-input-168-1aaa51b08a68>:3: DeprecationWarning:

The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.

<ipython-input-168-1aaa51b08a68>:3: DeprecationWarning:

The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.

<ipython-input-168-1aaa51b08a68>:3: DeprecationWarning:

The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.

<ipython-input-168-1aaa51b08a68>:3: DeprecationWarning:

The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.

<ipython-input-168-1aaa51b08a68>:3: DeprecationWarning:

The truth value of an empty array is ambigu

In [169]:
globalTd.head()

iyear  country         country_txt  region                   region_txt  \
0   1970       58  Dominican Republic       2  Central America & Caribbean   
1   1970      130              Mexico       1                North America   
2   1970      160         Philippines       5               Southeast Asia   
3   1970       78              Greece       8               Western Europe   
4   1970      101               Japan       4                    East Asia   

  provstate           city   latitude   longitude location  ...  \
0       NaN  Santo Domingo  18.456792  -69.951164      NaN  ...   
1   Federal    Mexico city  19.371887  -99.086624      NaN  ...   
2    Tarlac        Unknown  15.478598  120.599741      NaN  ...   
3    Attica         Athens  37.997490   23.762728      NaN  ...   
4   Fukouka        Fukouka  33.580412  130.396361      NaN  ...   

                  attacktype1_txt  targtype1                targtype1_txt  \
0                   Assassination         14  Private Citizens & Property   
1     Hostage Taking (Kidnapping)          7      Government (Diplomatic)   
2                   Assassination         10          Journalists & Media   
3               Bombing/Explosion          7      Government (Diplomatic)   
4  Facility/Infrastructure Attack          7      Government (Diplomatic)   

   targsubtype1                                   targsubtype1_txt  \
0          68.0                                     Named Civilian   
1          45.0  Diplomatic Personnel (outside of embassy, cons...   
2          54.0                    Radio Journalist/Staff/Facility   
3          46.0                                  Embassy/Consulate   
4          46.0                                  Embassy/Consulate   

                         corp1 weaptype1  weaptype1_txt nkill  nwound  
0                          NaN        13        Unknown   1.0     0.0  
1  Belgian Ambassador Daughter        13        Unknown   0.0     0.0  
2             Voice of America        13        Unknown   1.0     0.0  
3                          NaN         6     Explosives   NaN     NaN  
4                          NaN         8     Incendiary   NaN     NaN  

[5 rows x 26 columns]

All the columns that won't be needed in subsequent analysis have been eliminated.
Some columns had a significant amount of null entries, while others had the same data as other columns that had been erased.


## **PHASE - 4**


```
# DATA VISUALIZATION PART
```



### **PROBLEM: 1** Attacks overall by areas

In [170]:
# Code to calculate number of attacks by countries
totalAttacks = globalTd['country_txt'].value_counts().reset_index()

In [171]:
# Update the column names, which is easy to read the table
totalAttacks.rename(columns={'index':'country', 'country_txt':'total Attacks'}, inplace=True)

In [172]:
totalAttacks.head()

country  total Attacks
0         Iraq          24636
1     Pakistan          14368
2  Afghanistan          12731
3        India          11960
4     Colombia           8306

In [173]:
# Count the total number of attacks around the world
numberOfAttacks = totalAttacks['total Attacks'].sum()
print(numberOfAttacks)

181691


In [174]:
# Calculate the percentage (How many percent of the total number of attacks have occurred in any country?)
totalAttacks['prc'] = totalAttacks['total Attacks'].apply(lambda x: round((x/numberOfAttacks) *100, 2))

In [175]:
# Sorting the Data
totalAttacks.sort_values(by='country', inplace=True)

In [176]:
totalAttacks.head()

country  total Attacks   prc
2    Afghanistan          12731  7.01
97       Albania             80  0.04
17       Algeria           2743  1.51
203      Andorra              1  0.00
47        Angola            499  0.27



```
🔘 To better understand the given dataset, let's break it into three pieces.
1. Nations having the highest percentage (more than 1%) of terrorist attacks compared to other nations
2. Nations having a percentage of terrorist attacks between 1% and 0% compared to other nations
3. Nations with the fewest number of terrorist attacks compared to other nations (= 0).

```



In [177]:
import plotly.express as px
import pandas as pd

# Create a line plot using Plotly Express
line_fig = px.line(totalAttacks, x='country', y='prc', title='Line Plot of Percentage')
line_fig.update_traces(line=dict(color='blue'))

# Create a bar plot using Plotly Express
bar_fig = px.bar(totalAttacks, x='country', y='total Attacks', title='Bar Plot of Total Attacks')
bar_fig.update_traces(marker_color='green')

# Combine the plots using subplot
from plotly.subplots import make_subplots

fig = make_subplots(rows=1, cols=2, subplot_titles=('Percentage', 'Total Attacks'))

# Add the line and bar plots to the subplot
fig.add_trace(line_fig.data[0], row=1, col=1)
fig.add_trace(bar_fig.data[0], row=1, col=2)

# Set layout for subplot
fig.update_layout(width=1500, showlegend=False)

# Show the plot
fig.show()


### **PROBLEM: 2** Total assaults for the year, whether they went down or up.

In [178]:
# Fetch the data and count the total attacks per year
attacksPY = globalTd['iyear'].value_counts().reset_index()

In [179]:
# Rename the columns
attacksPY.rename(columns={'index':'year', 'iyear':'totalAttacks'}, inplace=True)

In [180]:
# Sort the table by the year
attacksPY.sort_values(by=['year'], inplace=True)

In [181]:
attacksPY.head()

year  totalAttacks
42  1970           651
46  1971           471
44  1972           568
45  1973           473
43  1974           581

In [182]:
attacksPY['prc'] = round((attacksPY['totalAttacks'] / numberOfAttacks) * 100, 2)

In [183]:
attacksPY.head()

year  totalAttacks   prc
42  1970           651  0.36
46  1971           471  0.26
44  1972           568  0.31
45  1973           473  0.26
43  1974           581  0.32

In [190]:
# ploting line plot to see the attacks over the years
fig = px.line(attacksPY, x = 'year', y='prc', hover_data={'year':True,
                                                                   'totalAttacks': True,
                                                                   'prc': True},
               labels={'year': 'Year', 'totalAttacks': 'Attacks', 'prc': 'Percentage'}, markers=True, line_shape='spline', color_discrete_sequence=["DarkRed"])

# Update layout (give xaxis title, yaxis title,etc...)
fig.update_layout(
    xaxis_title='Year',
    yaxis_title='Total Attacks',
    title='Analysis Attacks numbers per year whether decreased or increased',
    width=900
)

# Plot/ show the plot
fig.show()

### **PROBLEM: 3** The number of successful and unsuccessful assaults.

#### **STEP - 1** : SELECT TOTAL ATTACKS WITH GAP OF 3 YEARS

In [194]:
# We are gonna analysis attacks for every third year
selectedAttacks = attacksPY

In [195]:
selectedAttacks.reset_index(drop=True, inplace=True)

In [196]:
selectedAttacks

year  totalAttacks   prc
0   1970           651  0.36
1   1971           471  0.26
2   1972           568  0.31
3   1973           473  0.26
4   1974           581  0.32
5   1975           740  0.41
6   1976           923  0.51
7   1977          1319  0.73
8   1978          1526  0.84
9   1979          2662  1.47
10  1980          2662  1.47
11  1981          2586  1.42
12  1982          2544  1.40
13  1983          2870  1.58
14  1984          3495  1.92
15  1985          2915  1.60
16  1986          2860  1.57
17  1987          3183  1.75
18  1988          3721  2.05
19  1989          4324  2.38
20  1990          3887  2.14
21  1991          4683  2.58
22  1992          5071  2.79
23  1994          3456  1.90
24  1995          3081  1.70
25  1996          3058  1.68
26  1997          3197  1.76
27  1998           934  0.51
28  1999          1395  0.77
29  2000          1814  1.00
30  2001          1906  1.05
31  2002          1333  0.73
32  2003          1278  0.70
33  2004          1166  0.64
34  2005          2017  1.11
35  2006          2758  1.52
36  2007          3242  1.78
37  2008          4805  2.64
38  2009          4721  2.60
39  2010          4826  2.66
40  2011          5076  2.79
41  2012          8522  4.69
42  2013         12036  6.62
43  2014         16903  9.30
44  2015         14965  8.24
45  2016         13587  7.48
46  2017         10900  6.00

#### **STEP - 2**: COUNT THE TOTAL SUCCESSFULL ATTACKS AND SELECT THE DATA WITH THE 3 YEARS OF GAP

In [191]:
# Now count the unsuccessful attacks
successfullAttacks = globalTd.groupby('iyear')['success'].sum().reset_index()

In [192]:
# Rename the columns
successfullAttacks.rename(columns={'iyear':'year'}, inplace=True)

In [ ]:
# Now select the data with the gap of three years
successfullAttacks = successfullAttacks[::3]

In [ ]:
successfullAttacks.reset_index(drop=True, inplace=True)

In [193]:
successfullAttacks

year  success
0   1970      549
1   1971      420
2   1972      452
3   1973      433
4   1974      545
5   1975      705
6   1976      861
7   1977     1191
8   1978     1411
9   1979     2408
10  1980     2387
11  1981     2354
12  1982     2373
13  1983     2660
14  1984     3282
15  1985     2727
16  1986     2670
17  1987     2933
18  1988     3440
19  1989     4008
20  1990     3583
21  1991     4299
22  1992     4560
23  1994     3069
24  1995     2794
25  1996     2770
26  1997     2966
27  1998      859
28  1999     1237
29  2000     1637
30  2001     1689
31  2002     1213
32  2003     1149
33  2004     1080
34  2005     1910
35  2006     2660
36  2007     3123
37  2008     4402
38  2009     4430
39  2010     4385
40  2011     4606
41  2012     7600
42  2013    10484
43  2014    15015
44  2015    12676
45  2016    10975
46  2017     8652

#### **STEP - 3**: NOW COUNT THE UNSUCCESSFULL ATTACKS AND SELECT THE DATA WITH THE 3 YEARS OF GAP

In [197]:
# For count the unsuccessful attacks we can use below formula
# Unsuccessful = Total - successful
totalArray = np.array(selectedAttacks['totalAttacks'])
successArray = np.array(successfullAttacks['success'])

In [198]:
# Now unsuccessfull attacks
unsuccessArray = totalArray - successArray

In [199]:
unsuccessArray

array([ 102,   51,  116,   40,   36,   35,   62,  128,  115,  254,  275,
        232,  171,  210,  213,  188,  190,  250,  281,  316,  304,  384,
        511,  387,  287,  288,  231,   75,  158,  177,  217,  120,  129,
         86,  107,   98,  119,  403,  291,  441,  470,  922, 1552, 1888,
       2289, 2612, 2248])

In [200]:
# Make table of unsuccessArray
data = {
    'year': selectedAttacks['year'],
    'unsuccess': unsuccessArray
}

In [201]:
unsuccessAttacks = pd.DataFrame(data)

In [202]:
unsuccessAttacks

year  unsuccess
0   1970        102
1   1971         51
2   1972        116
3   1973         40
4   1974         36
5   1975         35
6   1976         62
7   1977        128
8   1978        115
9   1979        254
10  1980        275
11  1981        232
12  1982        171
13  1983        210
14  1984        213
15  1985        188
16  1986        190
17  1987        250
18  1988        281
19  1989        316
20  1990        304
21  1991        384
22  1992        511
23  1994        387
24  1995        287
25  1996        288
26  1997        231
27  1998         75
28  1999        158
29  2000        177
30  2001        217
31  2002        120
32  2003        129
33  2004         86
34  2005        107
35  2006         98
36  2007        119
37  2008        403
38  2009        291
39  2010        441
40  2011        470
41  2012        922
42  2013       1552
43  2014       1888
44  2015       2289
45  2016       2612
46  2017       2248

#### **STEP - 4**: NOW CONCATE TOTAL, SUCCESS & UNSUCCESSFULL TABLE

In [203]:
# First merge selectedAttacks and successfullAttacks
result = pd.merge(selectedAttacks, successfullAttacks, how='left', on='year')

In [204]:
# Now join unsuccessfull attacks
result = pd.merge(result, unsuccessAttacks, on='year')

In [205]:
result

year  totalAttacks   prc  success  unsuccess
0   1970           651  0.36      549        102
1   1971           471  0.26      420         51
2   1972           568  0.31      452        116
3   1973           473  0.26      433         40
4   1974           581  0.32      545         36
5   1975           740  0.41      705         35
6   1976           923  0.51      861         62
7   1977          1319  0.73     1191        128
8   1978          1526  0.84     1411        115
9   1979          2662  1.47     2408        254
10  1980          2662  1.47     2387        275
11  1981          2586  1.42     2354        232
12  1982          2544  1.40     2373        171
13  1983          2870  1.58     2660        210
14  1984          3495  1.92     3282        213
15  1985          2915  1.60     2727        188
16  1986          2860  1.57     2670        190
17  1987          3183  1.75     2933        250
18  1988          3721  2.05     3440        281
19  1989          4324  2.38     4008        316
20  1990          3887  2.14     3583        304
21  1991          4683  2.58     4299        384
22  1992          5071  2.79     4560        511
23  1994          3456  1.90     3069        387
24  1995          3081  1.70     2794        287
25  1996          3058  1.68     2770        288
26  1997          3197  1.76     2966        231
27  1998           934  0.51      859         75
28  1999          1395  0.77     1237        158
29  2000          1814  1.00     1637        177
30  2001          1906  1.05     1689        217
31  2002          1333  0.73     1213        120
32  2003          1278  0.70     1149        129
33  2004          1166  0.64     1080         86
34  2005          2017  1.11     1910        107
35  2006          2758  1.52     2660         98
36  2007          3242  1.78     3123        119
37  2008          4805  2.64     4402        403
38  2009          4721  2.60     4430        291
39  2010          4826  2.66     4385        441
40  2011          5076  2.79     4606        470
41  2012          8522  4.69     7600        922
42  2013         12036  6.62    10484       1552
43  2014         16903  9.30    15015       1888
44  2015         14965  8.24    12676       2289
45  2016         13587  7.48    10975       2612
46  2017         10900  6.00     8652       2248

#### **STEP-5**: NOW PLOT THE BAR CHART FOR BETTER COMPARISON

In [216]:
import pandas as pd
import plotly.express as px

# Create a stacked bar chart using Plotly Express
fig = px.bar(result, x='year', y=['unsuccess', 'success', 'totalAttacks'],
             title='Stacked Bar Chart: Total, Successful, and Unsuccessful Attacks by Country',
             labels={'value': '(attacks, succ, unsucc)', 'success': 'success', 'Unsuccess': 'unsuccess'},
             color_discrete_map={'totalAttacks': 'Black', 'success': 'Red', 'unsuccess': 'Green'})

# Show the plot
fig.show()


**INSIGHT**:Attacks have escalated over the past year, as is evident. Around 80 to 90% of all attacks were effective, leaving the remaining attacks to be unsuccessful. Over the years, assaults have been successful at a high rate around the world.

